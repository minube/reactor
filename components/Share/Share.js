import { func, string } from 'prop-types';
import React, { PureComponent } from 'react';
import { View } from 'react-native';

import { ENV, THEME } from '../../common';
import Button from '../Button';
import Dialog from '../Dialog';
import Icon from '../Icon';
import Text from '../Text';
import Touchable from '../Touchable';
import styles from './Share.style';

const { IS_WEB } = ENV;
const { COLOR } = THEME;
const shareNative = IS_WEB ? undefined : require('react-native').Share.share;

const OPTIONS = [
  { icon: 'facebookDark', title: 'Facebook', uri: 'https://www.facebook.com/sharer/sharer.php?u=' },
  { icon: 'twitterDark', title: 'Twitter', uri: 'http://www.twitter.com/share?url=' },
  { icon: 'emailDark', title: 'Email', uri: 'mailto:?&body=' },
  { icon: 'linkedinDark', title: 'Linkedin', uri: 'https://www.linkedin.com/shareArticle?mini=true&url=' },
  { icon: 'googleDark', title: 'Google Plus', uri: 'https://plus.google.com/share?url=' },
];

class Share extends PureComponent {
  static propTypes = {
    accentColor: string,
    caption: string,
    onCancel: func,
    onOption: func,
    onPress: func,
    title: string,
    uri: string.isRequired,
  };

  static defaultProps = {
    accentColor: COLOR.PRIMARY,
    caption: undefined,
    onCancel() {},
    onOption() {},
    onPress() {},
    title: 'Share',
  };

  constructor(props) {
    super(props);
    this.state = { visible: false };
  }

  _onCancel = () => {
    const { props: { onCancel } } = this;

    this.setState({ visible: false });
    onCancel();
  }

  _onOption = (baseUri) => {
    const { props: { onOption, uri } } = this;
    window.open(`${baseUri}${uri}`, '_target');
    onOption();
  }

  _onPress = () => {
    const {
      props: {
        caption, onPress, title, uri,
      },
    } = this;

    if (IS_WEB) {
      this.setState({ visible: true });
    } else {
      shareNative({ url: uri, title, message: caption });
    }
    onPress();
  }

  render() {
    const {
      _onCancel, _onOption, _onPress,
      props: { caption, accentColor, ...inherit },
      state: { visible },
    } = this;

    return (
      <View pointerEvents="auto">
        <Button {...inherit} icon="share" onPress={_onPress} />
        { IS_WEB
          && (
          <Dialog
            title={inherit.title}
            visible={visible}
            onClose={_onCancel}
            style={styles.dialog}
          >
            <Text lighten small>
              {caption}
            </Text>
            <View style={styles.options}>
              { OPTIONS.map(({ title, icon, uri }) => (
                <Touchable key={title} onPress={() => _onOption(uri)} style={styles.option}>
                  <Icon value={icon} style={styles.icon} />
                  <Text color={accentColor}>
                    {title}
                  </Text>
                </Touchable>))}
            </View>
          </Dialog>
          ) }
      </View>
    );
  }
}
export default Share;
