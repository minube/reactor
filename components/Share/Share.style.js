import { StyleSheet } from 'react-native';

import { STYLE, THEME } from '../../common';

const { COLOR, OFFSET, UNIT } = THEME;

export default StyleSheet.create({
  dialog: {
    minWidth: 'auto',
    width: UNIT * 36,
  },

  options: {
    borderTopColor: COLOR.BACKGROUND,
    borderTopWidth: 1,
    marginVertical: OFFSET,
  },

  option: {
    ...STYLE.ROW,
    paddingVertical: OFFSET,
    borderBottomColor: COLOR.BACKGROUND,
    borderBottomWidth: 1,
  },

  icon: {
    backgroundColor: COLOR.CONTENT,
    width: UNIT * 2,
    height: UNIT * 2,
    marginRight: UNIT,
  },
});
