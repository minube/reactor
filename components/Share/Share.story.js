import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { boolean, color, object, text } from '@storybook/addon-knobs/react';

import Share from './Share';

const LOREM = "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry standard dummy text ever since the 1500s.";
const STYLE = { borderRadius: 5, width: 128, height: 128 };

storiesOf('✅ Share', module)
  .addWithJSX('default', () => (
    <Share />
  ))
  .addWithJSX('title', () => (
    <Share title="Hello" />
  ))
  .addWithJSX('caption', () => (
    <Share caption={LOREM} />
  ))
  .addWithJSX('color', () => (
    <Share color="red" />
  ))
  .addWithJSX('⚡ onCancel (only web)', () => (
    <Share onCancel={action('Share.onCancel()')} />
  ))
  .addWithJSX('⚡ onOption (only web)', () => (
    <Share onOption={action('Share.onOption()')} />
  ))
  .addWithJSX('⚡ onPress', () => (
    <Share onPress={action('Share.onPress()')} />
  ))
  .addWithJSX('🏀 Playground', () => (
    <Share
      primary
      title={text('title', 'hello')}
      caption={text('caption', LOREM)}
      color={color('color', LOREM)}
      onCancel={action('Share.onCancel()')}
      onOption={action('Share.onOption()')}
      onPress={action('Share.onPress()')}
    />
  ));
