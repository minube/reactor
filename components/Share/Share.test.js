import React from 'react';
import renderer from 'react-test-renderer';

import Share from './Share';

const LOREM_Share = 'https://api.adorable.io/Shares/128/8';
const STYLE = { borderRadius: 5, width: 128, height: 128 };
const LOREM_URI = 'http://soyjavi.com';

describe('<Share>', () => {
  it('renders', () => {
    const tree = renderer.create(<Share uri={LOREM_URI} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {title}', () => {
    const tree = renderer.create(<share uri={LOREM_URI} title="Press me" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {caption}', () => {
    const tree = renderer.create(<share uri={LOREM_URI} caption="Hello world" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {accentColor}', () => {
    const tree = renderer.create(<share uri={LOREM_URI} accentColor="red" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('when {onCancel}', () => {
    // @TODO: Test <Share /> {onCancel}
  });

  it('when {onOption}', () => {
    // @TODO: Test <Share /> {onOption}
  });

  it('when {onPress}', () => {
    // @TODO: Test <Share /> {onPress}
  });
});

