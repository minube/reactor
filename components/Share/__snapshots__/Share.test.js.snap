// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Share> renders 1`] = `
<View
  pointerEvents="auto"
>
  <View
    accessible={true}
    onLayout={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    pointerEvents="box-only"
    style={
      Array [
        Object {
          "borderRadius": 3,
          "marginHorizontal": 5,
        },
        false,
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignContent": "center",
            "alignItems": "center",
            "borderRadius": 3,
            "justifyContent": "center",
          },
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexDirection": "row",
          },
          Object {
            "height": 45,
            "minWidth": 45,
            "paddingHorizontal": 30,
          },
          false,
          false,
          false,
          false,
          Object {
            "backgroundColor": "#9b9b9b",
          },
          false,
          false,
          false,
          false,
          false,
          false,
        ]
      }
    >
      <Image
        resizeMode="contain"
        source={
          Object {
            "testUri": "../../../components/Icon/assets/share.png",
          }
        }
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "height": 20,
              "width": 20,
            },
            undefined,
            undefined,
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "row",
            },
            Object {
              "marginLeft": 10,
            },
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#4a4a4a",
                "fontFamily": undefined,
              },
              false,
              Object {
                "fontWeight": "600",
              },
              false,
              false,
              Object {
                "fontSize": 16,
                "lineHeight": 20.8,
              },
              false,
              false,
              false,
              false,
              false,
              false,
              false,
              false,
              Object {
                "color": "#ffffff",
                "fontSize": 14,
              },
            ]
          }
        >
          Share
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderRadius": 3,
            "bottom": 0,
            "left": 0,
            "overflow": "hidden",
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#ffffff",
            "height": 0,
            "opacity": 0,
            "width": 0,
            "zIndex": 0,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`<Share> when {accentColor} 1`] = `
<share
  accentColor="red"
  uri="http://soyjavi.com"
/>
`;

exports[`<Share> when {caption} 1`] = `
<share
  caption="Hello world"
  uri="http://soyjavi.com"
/>
`;

exports[`<Share> when {title} 1`] = `
<share
  title="Press me"
  uri="http://soyjavi.com"
/>
`;
